draft-ietf-quic-transport.md

```diff
@@ -1207,41 +1208,82 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
-
-While endpoints select connection IDs as appropriate for their implementation,
-the connection ID MUST NOT include the unprotected sequence number.  Endpoints
-need to be able to recover the sequence number associated with each connection
-ID they generate without relying on information available to unaffiliated
-parties. A connection ID that encodes an unencrypted sequence number could be
-used to correlate connection IDs across network paths.
+An endpoint whose peer has selected a non-zero-length connection ID will
+potentially send packets using different connection IDs over the lifetime of a
+connection. These connection IDs are supplied by the peer using the
+NEW_CONNECTION_ID frame ({{frame-new-connection-id}}).
+
+
+### Issuing Connection IDs
+
+An endpoint issues connection IDs to its peer for the peer to use when sending
+packets. This allows the endpoint control over the strategy used to interpret
+connection IDs marked on packets that it receives.  These connection IDs are
+communicated to the peer using NEW_CONNECTION_ID frames
+({{frame-new-connection-id}}).
+
+The initial connection ID issued to the peer is sent by the endpoint as the
+Source Connection ID during the handshake, ensuring that the peer always begins
+the connection with at least one connection ID to use when sending.  When the
+endpoint receives the `requested_connection_ids` transport parameter during the
+handshake, it SHOULD supply its peer with the requested number of connection IDs
+via NEW_CONNECTION_ID frames.
+
+When an endpoint issues a connection ID, it MUST accept packets using this
+connection ID for the duration of the connection or until its peer invalidates
+the connection ID via a CONNECTION_ID_FINISHED frame
+({{frame-connection-id-finished}}).
+
+Implementations SHOULD ensure that peers have the requested number of connection
+IDs available to reduce the possibility of peers exhausting their supply of
+available connection IDs.  An implementation could do this by always supplying a
+new connection ID for each connection ID retired with a CONNECTION_ID_FINISHED
+frame.
+
+When a receiver of a packet notices that its peer is now using a previously
+unused connection ID, it MAY choose to supply its peer with a new connection ID
+using a NEW_CONNECTION_ID frame to reduce the possibility of its peer running
+out of available connection IDs.
+
+It its peer selected non-zero-length connection IDs, an endpoint that receives a
+packet with a previously unused connection ID, SHOULD switch to sending with a
+different connection ID from its set of valid connection IDs when sending
+responses.  This can help to ensure that when an endpoint migrates to a new path
+or changes connection ID on an existing path, the packets will use different
+connection IDs in both directions. Likewise, when an endpoint receives a
+CONNECTION_ID_FINISHED frame, it SHOULD retire any connection IDs that were used
+in packets generated in response to that connection ID.
```

Comment:
This is interesting, but risks the same synchronization problem.  Packets aren't necessarily "in response to" other packets in a QUIC sense.  Perhaps we turn it around and say "when the peer migrates to a new address, an endpoint SHOULD retire all CIDs which were used on packets directed to the previous address."  We should also retain the guidance that each CID only be used with a single remote address.

Reply:
Removed in response to 

Covering the rest of this in the summary below